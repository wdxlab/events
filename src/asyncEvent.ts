import { BaseEvent, EventOptions, Filter, Handler } from './baseEvent';
import { AggregateError } from './aggregateError';

export type AsyncHandler<TArg, TSender> = Handler<TArg, TSender, Promise<void>>;

export class AsyncEvent<TArg, TSender = void> extends BaseEvent<
  TArg,
  TSender,
  Promise<unknown[]>
> {
  constructor(options?: EventOptions<TArg, TSender, Promise<unknown[]>>) {
    super(options);

    this.sourceEvent?.on(async (arg, sender) => {
      const result = await this.emitInt(arg, sender);
      if (result.length) {
        throw new AggregateError(result);
      }
    });
  }

  async emit(arg: TArg, sender: TSender): Promise<unknown[]> {
    if (this.isReadOnly) {
      throw new Error('Event is readonly');
    }

    return this.emitInt(arg, sender);
  }

  protected async emitInt(arg: TArg, sender: TSender): Promise<unknown[]> {
    if (this.filter && !this.filter(arg, sender)) {
      return [];
    }

    const errors: unknown[] = [];

    for (const fn of this.subscribers) {
      try {
        await fn.call(sender, arg, sender);
      } catch (e) {
        errors.push(e);
        if (this.isBailMode) {
          break;
        }
      }
    }

    return errors;
  }

  for(filter: Filter<TArg, TSender>): AsyncEvent<TArg, TSender> {
    return new AsyncEvent<TArg, TSender>({ bail: this.isBailMode, filter, source: this });
  }
}
