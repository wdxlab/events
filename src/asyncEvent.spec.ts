import { AsyncEvent } from './asyncEvent';

it('should work', async () => {
  const event = new AsyncEvent<{ name: string }, null>();
  const handler = jest.fn(() => new Promise((r) => setTimeout(r, 500)));
  const handler2 = jest.fn(() => new Promise((r) => setTimeout(r, 200)));

  event.on(handler);
  event.on(handler2);

  await event.emit({ name: 'foo' }, null);
  expect(handler.mock.calls).toMatchSnapshot();
  expect(handler.mock.results).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  expect(handler2.mock.results).toMatchSnapshot();

  await event.emit({ name: 'bar' }, null);
  expect(handler.mock.calls).toMatchSnapshot();
  expect(handler.mock.results).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  expect(handler2.mock.results).toMatchSnapshot();
});

describe('error', () => {
  test('non-bail', async () => {
    const eventArg = { foo: 'hello' };
    const event = new AsyncEvent<{ foo: string }, void>();
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    event.on(handler1);
    event.on(handler2);
    event.on(handler3);

    const errors = await event.emit(eventArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(1);
  });

  test('bail', async () => {
    const eventArg = { foo: 'hello' };
    const event = new AsyncEvent<{ foo: string }, void>({ bail: true });
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    event.on(handler1);
    event.on(handler2);
    event.on(handler3);

    const errors = await event.emit(eventArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(0);
  });
});

test('readonly mode', async () => {
  const source = new AsyncEvent<number, void>();
  const target = new AsyncEvent<number, void>({ source });
  const handler1 = jest.fn();
  const handler2 = jest.fn(() => {
    throw new Error('Foo');
  });
  const handler3 = jest.fn();

  target.on(handler1);
  target.on(handler2);
  target.on(handler3);

  await expect(target.emit(123)).rejects.toThrow('Event is readonly');

  const errors = await source.emit(456);

  expect(errors).toHaveLength(1);
  expect(errors[0]).toBeInstanceOf(Error);
  expect(handler1).toBeCalledTimes(1);
  expect(handler2).toBeCalledTimes(1);
  expect(handler3).toBeCalledTimes(1);
});

it('filterable', async () => {
  const event = new AsyncEvent<{ name: string }, null>();
  const aaaFilter = event.for((arg) => arg.name === 'aaa');
  const bbbFilter = event.for((arg) => arg.name === 'bbb');
  const fooFilter = event.for((arg) => arg.name === 'foo');
  const handler1 = jest.fn();
  const handler2 = jest.fn();

  event.on(handler1);
  aaaFilter.on(handler2);
  bbbFilter.on(handler2);
  fooFilter.on(handler2);

  await event.emit({ name: 'ccc' }, null);
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  await event.emit({ name: 'ddd' }, null);
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  await event.emit({ name: 'foo' }, null);
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
});
