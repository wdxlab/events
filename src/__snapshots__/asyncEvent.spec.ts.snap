// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterable 1`] = `
[
  [
    {
      "name": "ccc",
    },
    null,
  ],
]
`;

exports[`filterable 2`] = `[]`;

exports[`filterable 3`] = `
[
  [
    {
      "name": "ccc",
    },
    null,
  ],
  [
    {
      "name": "ddd",
    },
    null,
  ],
]
`;

exports[`filterable 4`] = `[]`;

exports[`filterable 5`] = `
[
  [
    {
      "name": "ccc",
    },
    null,
  ],
  [
    {
      "name": "ddd",
    },
    null,
  ],
  [
    {
      "name": "foo",
    },
    null,
  ],
]
`;

exports[`filterable 6`] = `
[
  [
    {
      "name": "foo",
    },
    null,
  ],
]
`;

exports[`should work 1`] = `
[
  [
    {
      "name": "foo",
    },
    null,
  ],
]
`;

exports[`should work 2`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 3`] = `
[
  [
    {
      "name": "foo",
    },
    null,
  ],
]
`;

exports[`should work 4`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 5`] = `
[
  [
    {
      "name": "foo",
    },
    null,
  ],
  [
    {
      "name": "bar",
    },
    null,
  ],
]
`;

exports[`should work 6`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 7`] = `
[
  [
    {
      "name": "foo",
    },
    null,
  ],
  [
    {
      "name": "bar",
    },
    null,
  ],
]
`;

exports[`should work 8`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
  {
    "type": "return",
    "value": Promise {},
  },
]
`;
